### ISSUE TITLE
buildifier using bazel with release binaries

### ISSUE THREAD

**Issue by hanwen** [bazelbuild/buildtools#367]
I like the convenience of using bazel to version & download buildifier, but I'd rather not have our build depend on the Go toolchain. 

Would you be able to provide a rule that can be setup like 

  https://github.com/bazelbuild/buildtools/tree/master/buildifier#setup-and-usage-via-bazel

but downloads a binary release and runs that?

**Comment by pmbethe09** [comment]
So this is a
macro
buildifier_binary(version="0.15.0")
or something like that?  Which just needs a base buildtools in the WORKSPACE?

**Comment by hanwen** [comment]
yes; it should also automatically download for the host platform.

**Comment by aiuto** [comment]
Do you normally want specific version pinning, or is the more common case to simply to pin the buildifier binary to the matching Bazel release?

If the most common use case is the latter, we could solve this another way - simply have a prebuilt buildifier as part of the Bazel distribution. This goes for buildozer and any other Bazel command line tools. Because I fundamentally agree with you. When I get Bazel, I expect to have binaries for all the tools I need without having to build them (especially with toolchains I may not have).

Not that we should just bloat the distribution. We could package Bazel as two components.
1. bazel-core which is just bazel
2. bazel which depends on bazel-core but includes pre-built for everything under buildtools

**Comment by oferb** [comment]
Here's an example of how to bundle these tools with a repo:
https://github.com/google/startup-os/tree/master/tools/buildtools_wrappers
https://github.com/google/startup-os/blob/master/WORKSPACE#L67-L107

You can then call them the same way on both Linux and Mac:
https://github.com/google/startup-os/blob/master/commands.txt#L1-L8

**Comment by aiuto** [comment]
My latest thinking on this is that there will be two kinds of Bazel releases: "bazel" and "bazel-full". Alongside that, there will be binary releases for Buildifier.  At each Bazel release we will test against a specific buildifier release. The bazel packaging will stand alone. The bazel-full package will depend on the tested buildifier release.  If you install bazel-full, you get the tools at the needed versions, but you can still do an apt-get to get a newer buildifier release when it is available.  This still needs to be fleshed out for multiple platforms before it is a real plan, but the basic idea is that a large class of users should be able to just install bazel-full and not hunt for anything else.
